### PR TITLE
[CheckStyle Rule] invalid methods 

### LINKED ISSUE #6029: [FEATURE REQ] SpotBugs checks for runFinalizersOnExit
Tracking work item for additional SpotBugs static analysis checks

- java.lang.System.runFinalizersOnExit(boolean) => remove, this function is inherently unsafe.
- java.lang.Runtime.runFinalizersOnExit(boolean) =>remove, this function is inherently unsafe.
- java.lang.Math.random() => replace with java.security.SecureRandom.
- java.util.Random => replace with java.security.SecureRandom.

### PR DESCRIPTION
**Pull request by mssfang**
Tracking work item for additional CheckStyle static analysis checks

java.lang.System.runFinalizersOnExit(boolean) => remove, this function is inherently unsafe.
java.lang.Runtime.runFinalizersOnExit(boolean) =>remove, this function is inherently unsafe.
java.lang.Math.random() => replace with java.security.SecureRandom.
java.util.Random => replace with java.security.SecureRandom.

#6029

### CODE REVIEW

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/InvalidMethodsCheck.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Invalid method that should not be used in the Azure SDK. Such as it could avoid inherently unsafe method.
+ */
+public class InvalidMethodsCheck extends AbstractCheck {
```
Comment (srnagar) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/InvalidMethodsCheck.java]: `BlacklistedMethodsCheck`

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/InvalidMethodsCheck.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        }
+    }
+
+    private void invalidMethodCall(DetailAST methodCallToken) {
+        final String methodCallName = FullIdent.createFullIdentBelow(methodCallToken).getText();
+        // FullIdent.getText() will never return null but an empty string
+        if (methodCallName.isEmpty()) {
+            return;
+        }
+        for (String fullMethodName : methods) {
```
Comment (srnagar) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/InvalidMethodsCheck.java]: This for loop can be replaced by `methods.contains(fullMethodName)`.

File: eng/code-quality-reports/src/main/resources/checkstyle/checkstyle.xml [hunk 3]
```diff
@@ -378,6 +384,23 @@ page at http://checkstyle.sourceforge.net/config.html -->
     4) All classes should use ClientLogger as logger only but except ClientLogger itself -->
     <module name="com.azure.tools.checkstyle.checks.GoodLoggingCheck"/>
 
+    <!-- CUSTOM CHECKS -->
+    <!-- Invalid Method Checks:
+      Math.random method that should not be used in the Azure SDK. Use SecureRandom instead -->
+    <module name="com.azure.tools.checkstyle.checks.InvalidMethodsCheck">
+      <property name="methods" value="java.lang.Math.random, Math.random"/>
+      <property name="message" value="%s is not secure random. Use java.security.SecureRandom instead."/>
```
Comment (srnagar) [inline eng/code-quality-reports/src/main/resources/checkstyle/checkstyle.xml]:
```suggestion
      <property name="message" value="%s is not secure. Use java.security.SecureRandom instead."/>
```

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/BlacklistedMethodsCheck.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+/**
+ * Blacklist methods that should not be used in the Azure SDK. Such as it could avoid inherently unsafe method.
+ */
+public class BlacklistedMethodsCheck extends AbstractCheck {
+
+    /**
+     * Specified full name of invalid methods.
+     */
+    private final Set<String> methods = new HashSet<>(Arrays.asList());
```
Comment (alzimmermsft) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/BlacklistedMethodsCheck.java]: Why not use the base constructor?

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/BlacklistedMethodsCheck.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    private void blacklistedMethodsCheck(DetailAST methodCallToken) {
+        final String methodCallName = FullIdent.createFullIdentBelow(methodCallToken).getText();
+        // FullIdent.getText() will never return null but an empty string
+        if (methodCallName.isEmpty()) {
+            return;
+        }
+
+        if (methods.contains(methodCallName)) {
```
Comment (alzimmermsft) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/BlacklistedMethodsCheck.java]: What happens in the scenario of `import static Math.random()`? Is there any way to get class and method being called reflectively? Or is this an edge case we are willing to live with.

**Comment by JonathanGiles** [comment]
@mssfang are you going to update this PR so that it can be merged?

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/BlacklistedMethodsCheck.java (hunk 5 above)
Comment (mssfang) [reply eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/BlacklistedMethodsCheck.java]: What you get from the Token will be text based not the original object. So I believe the reflection won't work here. Please correct me if I am wrong.

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/BlacklistedMethodsCheck.java (hunk 4 above)
Comment (mssfang) [reply eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/BlacklistedMethodsCheck.java]: Will change to base constructor

**Comment by srnagar** [comment]
@mssfang Could you please take a look at this PR and see if this needs an update or if this should be closed?

**Comment by mssfang** [comment]
Sure. It needs to be updated. I will take a look at this PR.

**Comment by mssfang** [comment]
closing this PR since it been too long on opening status

**Comment by mssfang** [comment]
My bad. It should be better to just update the PR rather than closing it.

**Comment by JonathanGiles** [comment]
This PR is so old that I suspect it would be quicker to copy / paste the changes into a new PR. @mssfang Were you interested in doing this?

**Comment by mssfang** [comment]
Close this PR due to it is too old.
Moved the changes to this new PR: https://github.com/Azure/azure-sdk-for-java/pull/15092

**Pull request closed without merging**